app/models/server.rb

```diff
@@ -27,4 +32,104 @@ class Server < ApplicationRecord
     with: /([0-9A-Fa-f]{2}[:-]){5}([0-9A-Fa-f]{2})/,
     message: 'is not a valid MAC address'
   }
+
+  after_commit :save_users
+  after_initialize :assign_defaults
+
+  attr_writer :sudo_users
+  attr_writer :users
+
+  def user_ids=(new_user_ids)
+    @users = User.find(new_user_ids)
+  end
+
+  def sudo_user_ids=(new_sudo_users)
+    @sudo_users = User.find(new_sudo_users)
+  end
+
+  # in this model, the groups: users, sudo_users, and responsible_users shall be seperate groups
+  # They will be merged for writing puppet scripts and seperated when reading them
+
+  def sudo_users
+    read_users if @sudo_users.nil?
+
+    @sudo_users
+  end
+
+  def users
+    read_users if @users.nil?
+
+    @users
+  end
+
+  def all_users
+    (users + sudo_users + responsible_users).uniq
+  end
+
+  # There is actually only ever one responsible user
+  # This method is a convenience accessor to allow for easier array concatination
+  # It is also recommended to use this method instead of :responsible because we might want to allow multiple responsible users in the future.
+  def responsible_users
+    responsible.nil? ? [] : [responsible]
+  end
+
+  def assign_defaults
+    self.name = '' if name.nil?
+  end
+
+  private
+
+  def read_users
+    begin
+      GitHelper.open_git_repository do
+        remote_users = Puppetscript.read_node_file name
+        @users = remote_users[:users] || [] if @users.nil?
+        @sudo_users = remote_users[:admins] || [] if @sudo_users.nil?
+      end
+    rescue Git::GitExecuteError, RuntimeError => e
+      Rails.logger.error(e)
+      # ensure users and sudo_users are always valid arrays
+      @users = [] if @users.nil?
+      @sudo_users = [] if @sudo_users.nil?
+    end
+    @users -= @sudo_users
+    @sudo_users -= responsible_users
+  end
+
+  def convert_to_user(user_or_id)
+    user.is_a? User ? user : User.find(user_or_id)
+  end
+
+  def commit_message(git_writer)
```

Comment:
This exact code is also found in `app/models/virtual_machine_config.rb` and `app/models/request.rb`. Maybe it could be be moved to a shared helper?